### PR TITLE
IA-1943: Backend - creation of user roles feature

### PR DESCRIPTION
**Pull request by hakifran**
Explain what problem this PR is resolving
- There was no user role management in IASO
- This feature is for managing user roles(add, update, delete and list user role)
Related JIRA tickets : [IA-1943](https://bluesquare.atlassian.net/browse/IA-1943)

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [ ] Are my typescript files well typed
- [ ] New translations have been added or updated if new strings have been introduced in the frontend
- [x] My migrations file are included
- [x] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes
- Added UserRole model to be able to register created_at and updated_at for user groups
- Added defaults user_roles and their respective permissions from migrations
- Added create, partial_update, delete and list for user roles in the rest APIs calls
- Added unit test related to APIs calls added

## How to test
- Create:
      - Link: `'/api/userroles/'`
      - Payload: `{"name": "New user role name", "permissions": [list of permissions code names]}`
      - Method: POST
- list: 
      - Link: `'/api/userroles/'`
      - Payload: `{"search": "Name"}`
      - Method: GET
- partial_update: 
      - Link: `'/api/userroles/{userrole_id}'`
      - Payload: `{"name": "Modified name",  "permissions": [list of new permissions to be added]}`
      - Method: PATCH
- retrieve: 
      - Link: `'/api/userroles/{userrole_id}'`
      - Method: GET
- delete: 
      - Link: `'/api/userroles/{userrole_id}'`
      - Method: DELETE

[IA-1943]: https://bluesquare.atlassian.net/browse/IA-1943?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by olethanh** [comment]
@madewulf Correct me if I'm wrong, but it seems to me that User Role should be per tenant? Especially if customer can modify them, otherhise it open security problem.
I'm removing the Release tag since this is quite blocking and since we don't have a Frontend it is not urgent to merge it anyway

Also this is not really blocking but I'm a bit surprised that this is code is done in the old style and not the new one with serializer  and ModelView as described in the Guidelines https://wiki.bluesquare.org/en/Dev-team/product-management/iaso/backend-guidelines

**Review by olethanh — CHANGES_REQUESTED** [review]
see previous comment, for me this need  tenancy

File: iaso/api/user_roles.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            group.permissions.clear()
+            for permission_codename in permissions:
+                permission = get_object_or_404(Permission, codename=permission_codename)
+                group.permissions.add(permission)
+        group.save()
+        user_role.save()
+        return user_role
+
+
+class HasRolesPermission(permissions.BasePermission):
```
Comment (mathvdh) [inline iaso/api/user_roles.py]: See comment below, you can remove this code.

File: iaso/api/user_roles.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    GET /api/userroles/
+    GET /api/userroles/<id>
+    UPDATE /api/userroles/<id>
+    DELETE /api/userroles/<id>
+    """
+
+    # FIXME : replace by a model viewset
+
+    permission_classes = [permissions.IsAuthenticated, HasRolesPermission]
```
Comment (mathvdh) [inline iaso/api/user_roles.py]: You don't need to code `HasRolesPermission`, you can replace it by `HasPermission("menupermissions.iaso_user_roles")` (need to import `from iaso.api.common import HasPermission`)

File: iaso/api/user_roles.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    serializer_class = UserRoleSerializer
+    http_method_names = ["get", "post", "put", "delete"]
+
+    def get_queryset(self) -> QuerySet[UserRole]:
+        queryset = UserRole.objects.all()
```
Comment (mathvdh) [inline iaso/api/user_roles.py]: If it's multi-tenant and linked to an account, shouldn't the base be filtered by account ?
Using : `self.request.user.iaso_profile.account`

File: iaso/migrations/0211_auto_20230609_0846.py
Comment (mathvdh) [inline iaso/migrations/0211_auto_20230609_0846.py:1]: Does this python code to create the default groups need to be run on all deployments ??

**Review by mathvdh — CHANGES_REQUESTED** [review]
See comments above. Mainly small changes ...

File: iaso/migrations/0211_auto_20230609_0846.py
Comment (hakifran) [reply iaso/migrations/0211_auto_20230609_0846.py:1]: @mathvdh in the ticket https://bluesquare.atlassian.net/browse/IA-1943 there is a file in which Laure specify 5 user roles and their related permissions so I thought those are default ones. I just asked laure the question in the ticket.

Comment (mathvdh) [reply iaso/migrations/0211_auto_20230609_0846.py:1]: I don't know, I was just wondering .. maybe this can be confirmed by martin or laure ...

Comment (hakifran) [reply iaso/migrations/0211_auto_20230609_0846.py:1]: @mathvdh after a discussion with Laure we decided to remove the migration and it's done

File: iaso/api/user_roles.py (hunk 1 above)
Comment (hakifran) [reply iaso/api/user_roles.py]: Done

File: iaso/api/user_roles.py (hunk 2 above)
Comment (hakifran) [reply iaso/api/user_roles.py]: Done

**Review by mathvdh — APPROVED** [review]
Top

**Comment by hakifran** [comment]
> see previous comment, for me this need tenancy

This has been done

**Pull request merged**